src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -3799,6 +3805,159 @@ private bool SelfDescribingEvents
+                etwProvider.Register(id, name);
+#if TARGET_WINDOWS
+                byte[] providerMetadata = Statics.MetadataForString(name, 0, 0, 0);
+                fixed (byte* pMetadata = providerMetadata)
+                {
+                    etwProvider.SetInformation(
+                        Interop.Advapi32.EVENT_INFO_CLASS.SetTraits,
+                        pMetadata,
+                        (uint)providerMetadata.Length);
+                }
+#endif // TARGET_WINDOWS
+                lock (s_preregisteredEtwProviders)
+                {
+                    s_preregisteredEtwProviders[id] = etwProvider;
+                }
+
+#if FEATURE_PERFTRACING
+                OverrideEventProvider eventPipeProvider = new OverrideEventProvider(eventSourceFactory, EventProviderType.EventPipe);
+                eventPipeProvider.Register(id, name);
+                lock (s_preregisteredEventPipeProviders)
+                {
+                    s_preregisteredEventPipeProviders[name] = eventPipeProvider;
+                }
+#endif
+            }
+            catch (Exception)
+            {
+                // If there is a failure registering then the normal EventSource.Initialize() path can try to register
+                // again if/when the EventSource is constructed.
+            }
+        }
+
+        internal static void EnsurePreregisteredEventSourcesExist()
+        {
+            if (!EventSource.IsSupported)
+            {
+                return;
+            }
+
+            // In a multi-threaded race its possible that one thread will be creating the EventSources while a 2nd thread
+            // exits this function and observes the s_EventSources list without the new EventSources in it.
+            // There is no known issue here having a small window of time where the pre-registered EventSources are not in
+            // the list as long as we still guarantee they get initialized in the near future and reported to the
+            // same EventListener.OnEventSourceCreated() callback.
+            Func<EventSource?>[] factories;
+            lock(s_preregisteredEventSourceFactories)
+            {
+                factories = s_preregisteredEventSourceFactories.ToArray();
+                s_preregisteredEventSourceFactories.Clear();
+            }
+            foreach (Func<EventSource?> factory in factories)
+            {
+                factory();
+            }
+        }
+
+        private static List<Func<EventSource?>> s_preregisteredEventSourceFactories = new List<Func<EventSource?>>();
+
+        private static OverrideEventProvider? TryGetPreregisteredEtwProvider(Guid id)
+        {
+            lock (s_preregisteredEtwProviders)
+            {
+                s_preregisteredEtwProviders.TryGetValue(id, out OverrideEventProvider? provider);
+                s_preregisteredEtwProviders.Remove(id);
```

Comment:
should it be 

```C#
if (s_preregisteredEtwProviders.TryGetValue(id, out OverrideEventProvider? provider))
{
                s_preregisteredEtwProviders.Remove(id);
}
```

Reply:
Use Remove overload that returns the value that was removed?

`s_preregisteredEtwProviders.Remove(id, out OverrideEventProvider? provider);`

Reply:
didn't realize that overload existed, thanks!